mo2_cpp.cmake

```diff
@@ -0,0 +1,289 @@
+
+	# VS emits a warning for LTCG, at least for uibase, so maybe not required?
+	set(OPTIMIZE_LINK_FLAGS "/LTCG /INCREMENTAL:NO /OPT:REF /OPT:ICF")
+
+	if(${MO2_WARNINGS})
+		set(COMPILE_FLAGS "${COMPILE_FLAGS} /Wall /wd4464")
+	endif()
+
+	if(NOT ${MO2_PERMISSIVE})
+		set(COMPILE_FLAGS "${COMPILE_FLAGS} /permissive-")
+	endif()
+
+	if(${MO2_BIGOBJ})
+		set(COMPILE_FLAGS "${COMPILE_FLAGS} /bigobj")
+	endif()
+
+	# find source files
+	if(DEFINED AUTOGEN_BUILD_DIR)
+		set(UI_HEADERS_DIR ${AUTOGEN_BUILD_DIR})
+	else()
+		set(UI_HEADERS_DIR ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}_autogen/include_RelWithDebInfo)
+	endif()
+
+	file(GLOB_RECURSE source_files CONFIGURE_DEPENDS *.cpp)
+	file(GLOB_RECURSE header_files CONFIGURE_DEPENDS *.h)
+	file(GLOB_RECURSE qrc_files CONFIGURE_DEPENDS *.qrc)
+	file(GLOB_RECURSE rc_files CONFIGURE_DEPENDS *.rc)
+	file(GLOB_RECURSE ui_files CONFIGURE_DEPENDS *.ui)
+	file(GLOB_RECURSE ui_header_files CONFIGURE_DEPENDS ${UI_HEADERS_DIR}/*.h)
+	file(GLOB_RECURSE rule_files CONFIGURE_DEPENDS ${CMAKE_BINARY_DIR}/*.rule)
+	file(GLOB_RECURSE misc_files CONFIGURE_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/../*.natvis)
+
+	source_group(src REGULAR_EXPRESSION ".*\\.(h|cpp)")
+	source_group(ui REGULAR_EXPRESSION ".*\\.ui")
+	source_group(cmake FILES CMakeLists.txt)
+	source_group(autogen FILES ${rule_files} ${qm_files} ${ui_header_files})
+	source_group(autogen REGULAR_EXPRESSION ".*\\cmake_pch.*")
+	source_group(resources FILES ${rc_files} ${qrc_files})
+
+	if(${MO2_TRANSLATIONS})
+		find_package(Qt5LinguistTools)
+		qt5_create_translation(
+			qm_files
+			${source_files} ${header_files} ${ui_files} ${MO2_EXTRA_TRANSLATIONS}
+			${CMAKE_CURRENT_SOURCE_DIR}/${PROJECT_NAME}_en.ts
+			OPTIONS -silent
+		)
+	endif()
+
+	target_sources(${MO2_TARGET}
+		PRIVATE
+		${source_files}
+		${header_files}
+		${ui_files}
+		${ui_header_files}
+		${qrc_files}
+		${rc_files}
+		${misc_files}
+		${qm_files})
+
+	execute_process(
+	  COMMAND git log -1 --format=%h
+	  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+	  OUTPUT_VARIABLE GIT_COMMIT_HASH
+	  OUTPUT_STRIP_TRAILING_WHITESPACE
+	)
+
+	target_compile_definitions(
+		${MO2_TARGET}
+		PRIVATE
+		_UNICODE
+		UNICODE
+		NOMINMAX
+		_CRT_SECURE_NO_WARNINGS
+		BOOST_CONFIG_SUPPRESS_OUTDATED_MESSAGE
+		_SILENCE_CXX17_CODECVT_HEADER_DEPRECATION_WARNING
+		QT_MESSAGELOGCONTEXT
+		GITID="${GIT_COMMIT_HASH}")
+
+	if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/pch.h)
+		target_precompile_headers(${PROJECT_NAME}
+			PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/pch.h)
+	endif()
+
+    if(${MO2_CLI})
+        if (CMAKE_GENERATOR MATCHES "Visual Studio")
+            set_target_properties(${MO2_TARGET} PROPERTIES
+                COMMON_LANGUAGE_RUNTIME "")
+        else()
+            set(COMPILE_FLAGS "${COMPILE_FLAGS} /clr")
+            string(REPLACE "/EHs" "/EHa" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+        endif()
+    endif()
+
+	set_target_properties(${MO2_TARGET} PROPERTIES
+		COMPILE_FLAGS "${COMPILE_FLAGS}"
```

Comment:
It's probably a better idea to use `target_compile_options` instead of appending to a variable and setting the property to the value of that variable. I *think* the way it's done here might break CMake options that implicitly set compile flags, e.g. the `COMMON_LANGUAGE_RUNTIME` target property sets the `/clr` flag, but this *might* overwrite it depending on the order they happen in and how friendly CMake feels on a given day. It'll *definitely* overwrite anything that was set directory-wide before the target was declared.

Reply:
Thanks, done for standard compile flags but I'm not sure I can do it for `COMPILE_FLAGS_RELWITHDEBINFO` and `LINK_FLAGS_RELWITHDEBINFO`...

Reply:
In newer versions of CMake, `COMPILE_FLAGS` isn't a thing any more, and is emulated via `COMPILE_OPTIONS` and there isn't a version of that property for each config - if you use `COMPILE_FLAGS_RELWITHDEBINFO` I'm under the impression it adds it to `COMPILE_OPTIONS` with a generator expression that means it only shows up for the relevant config. I think they recommend you use `target_compile_options` with a generator expression instead.

Reply:
I used generator expressions and it seems to work, so, yay.